File: app/src/main/java/dev/ragnarok/fenrir/mvp/presenter/AudiosLocalPresenter.java
package dev.ragnarok.fenrir.mvp.presenter;

import static dev.ragnarok.fenrir.Includes.provideMainThreadScheduler;
import static dev.ragnarok.fenrir.util.Utils.findIndexById;

import android.content.Context;
import android.net.Uri;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import dev.ragnarok.fenrir.Includes;
import dev.ragnarok.fenrir.R;
import dev.ragnarok.fenrir.db.Stores;
import dev.ragnarok.fenrir.media.music.MusicPlaybackService;
import dev.ragnarok.fenrir.model.Audio;
import dev.ragnarok.fenrir.mvp.presenter.base.AccountDependencyPresenter;
import dev.ragnarok.fenrir.mvp.view.IAudiosLocalView;
import dev.ragnarok.fenrir.place.PlaceFactory;
import dev.ragnarok.fenrir.settings.Settings;
import dev.ragnarok.fenrir.upload.IUploadManager;
import dev.ragnarok.fenrir.upload.Upload;
import dev.ragnarok.fenrir.upload.UploadDestination;
import dev.ragnarok.fenrir.upload.UploadIntent;
import dev.ragnarok.fenrir.upload.UploadResult;
import dev.ragnarok.fenrir.util.Pair;
import dev.ragnarok.fenrir.util.RxUtils;
import dev.ragnarok.fenrir.util.Utils;
import io.reactivex.rxjava3.disposables.CompositeDisposable;

public class AudiosLocalPresenter extends AccountDependencyPresenter<IAudiosLocalView> {

    private final ArrayList<Audio> origin_audios;
    private final ArrayList<Audio> audios;
    private final CompositeDisposable audioListDisposable = new CompositeDisposable();
    private final IUploadManager uploadManager;
    private final List<Upload> uploadsData;
    private final UploadDestination destination;
    private final UploadDestination remotePlay;
    private boolean actualReceived;
    private boolean loadingNow;
    private String query;
    private boolean errorPermissions;
    private boolean doAudioLoadTabs;
    private int bucket_id;

    public AudiosLocalPresenter(int accountId, @Nullable Bundle savedInstanceState) {
        super(accountId, savedInstanceState);
        destination = UploadDestination.forAudio(accountId);
        remotePlay = UploadDestination.forRemotePlay();
        uploadManager = Includes.getUploadManager();
        uploadsData = new ArrayList<>(0);
        audios = new ArrayList<>();
        origin_audios = new ArrayList<>();
    }

    public void fireBucketSelected(int bucket_id) {
        this.bucket_id = bucket_id;
        fireRefresh();
    }

    public void firePrepared() {
        appendDisposable(uploadManager.get(getAccountId(), destination)
                .compose(RxUtils.applySingleIOToMainSchedulers())
                .subscribe(this::onUploadsDataReceived));

        appendDisposable(uploadManager.observeAdding()
                .observeOn(provideMainThreadScheduler())
                .subscribe(this::onUploadsAdded));

        appendDisposable(uploadManager.observeDeleting(true)
                .observeOn(provideMainThreadScheduler())
                .subscribe(this::onUploadDeleted));

        appendDisposable(uploadManager.observeResults()
                .filter(pair -> destination.compareTo(pair.getFirst().getDestination()) || remotePlay.compareTo(pair.getFirst().getDestination()))
                .observeOn(provideMainThreadScheduler())
                .subscribe(this::onUploadResults));

        appendDisposable(uploadManager.obseveStatus()
                .observeOn(provideMainThreadScheduler())
                .subscribe(this::onUploadStatusUpdate));

        appendDisposable(uploadManager.observeProgress()
                .observeOn(provideMainThreadScheduler())
                .subscribe(this::onProgressUpdates));

        fireRefresh();
    }

    public void setLoadingNow(boolean loadingNow) {
        this.loadingNow = loadingNow;
        resolveRefreshingView();
    }

    private boolean checkTittleArtists(@NonNull Audio data, @NonNull String q) {
        String[] r = q.split("( - )|( )|( {2})", 2);
        if (r.length >= 2) {
            return Utils.safeCheck(data.getArtist(), () -> data.getArtist().toLowerCase().contains(r[0].toLowerCase()))
                    && Utils.safeCheck(data.getTitle(), () -> data.getTitle().toLowerCase().contains(r[1].toLowerCase()));
        }
        return false;
    }

    public void updateCriteria() {
        setLoadingNow(true);
        audios.clear();
        if (Utils.isEmpty(query)) {
            audios.addAll(origin_audios);
            setLoadingNow(false);
            callView(IAudiosLocalView::notifyListChanged);
            return;
        }
        for (Audio i : origin_audios) {
            if (Utils.safeCheck(i.getTitle(), () -> i.getTitle().toLowerCase().contains(query.toLowerCase()))
                    || Utils.safeCheck(i.getArtist(), () -> i.getArtist().toLowerCase().contains(query.toLowerCase())) || checkTittleArtists(i, query)) {
                audios.add(i);
            }
        }
        setLoadingNow(false);
        callView(IAudiosLocalView::notifyListChanged);
    }

    public void fireQuery(String q) {
        if (Utils.isEmpty(q))
            query = null;
        else {
            query = q;
        }
        updateCriteria();
    }

    @Override
    public void onGuiResumed() {
        super.onGuiResumed();
        resolveRefreshingView();
        if (doAudioLoadTabs) {
            return;
        } else {
            doAudioLoadTabs = true;
        }
        callView(IAudiosLocalView::checkPermission);
    }

    private void resolveRefreshingView() {
        callResumedView(v -> v.displayRefreshing(loadingNow));
    }

    public void requestList() {
        setLoadingNow(true);
        if (bucket_id == 0) {
            audioListDisposable.add(Stores.getInstance()
                    .localMedia()
                    .getAudios(getAccountId())
                    .compose(RxUtils.applySingleIOToMainSchedulers())
                    .subscribe(this::onListReceived, this::onListGetError));
        } else {
            audioListDisposable.add(Stores.getInstance()
                    .localMedia()
                    .getAudios(getAccountId(), bucket_id)
                    .compose(RxUtils.applySingleIOToMainSchedulers())
                    .subscribe(this::onListReceived, this::onListGetError));
        }
    }

    private void onListReceived(List<Audio> data) {
        if (Utils.isEmpty(data)) {
            actualReceived = true;
            setLoadingNow(false);
            return;
        }
        origin_audios.clear();
        actualReceived = true;
        origin_audios.addAll(data);
        updateCriteria();
        setLoadingNow(false);
    }

    public void playAudio(Context context, int position) {
        MusicPlaybackService.startForPlayList(context, audios, position, false);
        if (!Settings.get().other().isShow_mini_player())
            PlaceFactory.getPlayerPlace(getAccountId()).tryOpenWith(context);
    }

    public void fireDelete(int position) {
        audios.remove(position);
        callView(v -> v.notifyItemRemoved(position));
    }

    @Override
    public void onDestroyed() {
        audioListDisposable.dispose();
        super.onDestroyed();
    }

    public void fireRemoveClick(Upload upload) {
        uploadManager.cancel(upload.getId());
    }

    private void onListGetError(Throwable t) {
        setLoadingNow(false);
        callResumedView(v -> showError(v, Utils.getCauseIfRuntime(t)));
    }

    public void fireFileForUploadSelected(String file) {
        UploadIntent intent = new UploadIntent(getAccountId(), destination)
                .setAutoCommit(true)
                .setFileUri(Uri.parse(file));

        uploadManager.enqueue(Collections.singletonList(intent));
    }

    public void fireFileForRemotePlaySelected(String file) {
        UploadIntent intent = new UploadIntent(getAccountId(), remotePlay)
                .setAutoCommit(true)
                .setFileUri(Uri.parse(file));

        uploadManager.enqueue(Collections.singletonList(intent));
    }

    public int getAudioPos(Audio audio) {
        if (!Utils.isEmpty(audios) && audio != null) {
            int pos = 0;
            for (Audio i : audios) {
                if (i.getId() == audio.getId() && i.getOwnerId() == audio.getOwnerId()) {
                    i.setAnimationNow(true);
                    int finalPos = pos;
                    callView(v -> v.notifyItemChanged(finalPos));
                    return pos;
                }
                pos++;
            }
        }
        return -1;
    }

    public void firePermissionsCanceled() {
        errorPermissions = true;
    }

    public void fireRefresh() {
        if (errorPermissions) {
            errorPermissions = false;
            callView(IAudiosLocalView::checkPermission);
            return;
        }
        audioListDisposable.clear();
        requestList();
    }

    public void fireScrollToEnd() {
        if (actualReceived) {
            requestList();
        }
    }

    @Override
    public void onGuiCreated(@NonNull IAudiosLocalView view) {
        super.onGuiCreated(view);
        view.displayList(audios);
        view.displayUploads(uploadsData);
        resolveUploadDataVisibility();
    }

    private void onUploadsDataReceived(List<Upload> data) {
        uploadsData.clear();
        uploadsData.addAll(data);

        resolveUploadDataVisibility();
    }

    private void onUploadResults(Pair<Upload, UploadResult<?>> pair) {
        Audio obj = (Audio) pair.getSecond().getResult();
        if (obj.getId() == 0)
            callView(v -> v.getCustomToast().showToastError(R.string.error));
        else {
            callView(v -> v.getCustomToast().showToast(R.string.uploaded));
        }

    }

    private void onProgressUpdates(List<IUploadManager.IProgressUpdate> updates) {
        for (IUploadManager.IProgressUpdate update : updates) {
            int index = findIndexById(uploadsData, update.getId());
            if (index != -1) {
                callView(view -> view.notifyUploadProgressChanged(index, update.getProgress(), true));
            }
        }
    }

    private void onUploadStatusUpdate(Upload upload) {
        int index = findIndexById(uploadsData, upload.getId());
        if (index != -1) {
            callView(view -> view.notifyUploadItemChanged(index));
        }
    }

    private void onUploadsAdded(List<Upload> added) {
        for (Upload u : added) {
            if (destination.compareTo(u.getDestination())) {
                int index = uploadsData.size();
                uploadsData.add(u);
                callView(view -> view.notifyUploadItemsAdded(index, 1));
            }
        }

        resolveUploadDataVisibility();
    }

    private void onUploadDeleted(int[] ids) {
        for (int id : ids) {
            int index = findIndexById(uploadsData, id);
            if (index != -1) {
                uploadsData.remove(index);
                callView(view -> view.notifyUploadItemRemoved(index));
            }
        }

        resolveUploadDataVisibility();
    }

    private void resolveUploadDataVisibility() {
        callView(v -> v.setUploadDataVisible(!uploadsData.isEmpty()));
    }

}

File: app/src/main/java/dev/ragnarok/fenrir/mvp/presenter/AudiosLocalPresenter.kt
package dev.ragnarok.fenrir.mvp.presenter

import android.content.Context
import android.net.Uri
import android.os.Bundle
import dev.ragnarok.fenrir.Includes
import dev.ragnarok.fenrir.Includes.provideMainThreadScheduler
import dev.ragnarok.fenrir.R
import dev.ragnarok.fenrir.db.Stores
import dev.ragnarok.fenrir.fromIOToMain
import dev.ragnarok.fenrir.media.music.MusicPlaybackService.Companion.startForPlayList
import dev.ragnarok.fenrir.model.Audio
import dev.ragnarok.fenrir.mvp.presenter.base.AccountDependencyPresenter
import dev.ragnarok.fenrir.mvp.view.IAudiosLocalView
import dev.ragnarok.fenrir.place.PlaceFactory.getPlayerPlace
import dev.ragnarok.fenrir.settings.Settings
import dev.ragnarok.fenrir.upload.*
import dev.ragnarok.fenrir.upload.IUploadManager.IProgressUpdate
import dev.ragnarok.fenrir.upload.UploadDestination.Companion.forAudio
import dev.ragnarok.fenrir.upload.UploadDestination.Companion.forRemotePlay
import dev.ragnarok.fenrir.util.Pair
import dev.ragnarok.fenrir.util.Utils.SafeCallCheckInt
import dev.ragnarok.fenrir.util.Utils.findIndexById
import dev.ragnarok.fenrir.util.Utils.getCauseIfRuntime
import dev.ragnarok.fenrir.util.Utils.safeCheck
import io.reactivex.rxjava3.disposables.CompositeDisposable
import java.util.*

class AudiosLocalPresenter(accountId: Int, savedInstanceState: Bundle?) :
    AccountDependencyPresenter<IAudiosLocalView>(accountId, savedInstanceState) {
    private val origin_audios: ArrayList<Audio>
    private val audios: ArrayList<Audio>
    private val audioListDisposable = CompositeDisposable()
    private val uploadManager: IUploadManager = Includes.uploadManager
    private val uploadsData: MutableList<Upload>
    private val destination: UploadDestination = forAudio(accountId)
    private val remotePlay: UploadDestination = forRemotePlay()
    private var actualReceived = false
    private var loadingNow = false
    private var query: String? = null
    private var errorPermissions = false
    private var doAudioLoadTabs = false
    private var bucket_id = 0
    fun fireBucketSelected(bucket_id: Int) {
        this.bucket_id = bucket_id
        fireRefresh()
    }

    fun firePrepared() {
        appendDisposable(uploadManager[accountId, destination]
            .fromIOToMain()
            .subscribe { data -> onUploadsDataReceived(data) })
        appendDisposable(uploadManager.observeAdding()
            .observeOn(provideMainThreadScheduler())
            .subscribe { added -> onUploadsAdded(added) })
        appendDisposable(uploadManager.observeDeleting(true)
            .observeOn(provideMainThreadScheduler())
            .subscribe { ids -> onUploadDeleted(ids) })
        appendDisposable(uploadManager.observeResults()
            .filter {
                destination.compareTo(
                    it.first.destination
                ) || remotePlay.compareTo(it.first.destination)
            }
            .observeOn(provideMainThreadScheduler())
            .subscribe { pair -> onUploadResults(pair) })
        appendDisposable(uploadManager.obseveStatus()
            .observeOn(provideMainThreadScheduler())
            .subscribe { upload -> onUploadStatusUpdate(upload) })
        appendDisposable(uploadManager.observeProgress()
            .observeOn(provideMainThreadScheduler())
            .subscribe { updates -> onProgressUpdates(updates) })
        fireRefresh()
    }

    fun setLoadingNow(loadingNow: Boolean) {
        this.loadingNow = loadingNow
        resolveRefreshingView()
    }

    private fun checkTittleArtists(data: Audio, q: String): Boolean {
        val r = q.split("( - )|( )|( {2})".toRegex(), 2).toTypedArray()
        return if (r.size >= 2) {
            (safeCheck(
                data.artist,
                object : SafeCallCheckInt {
                    override fun check(): Boolean {
                        return data.artist.lowercase(Locale.getDefault()).contains(
                            r[0].lowercase(
                                Locale.getDefault()
                            )
                        )
                    }
                })
                    && safeCheck(
                data.title,
                object : SafeCallCheckInt {
                    override fun check(): Boolean {
                        return data.title.lowercase(Locale.getDefault()).contains(
                            r[1].lowercase(
                                Locale.getDefault()
                            )
                        )
                    }
                }))
        } else false
    }

    private fun updateCriteria() {
        setLoadingNow(true)
        audios.clear()
        if (query.isNullOrEmpty()) {
            audios.addAll(origin_audios)
            setLoadingNow(false)
            view?.notifyListChanged()
            return
        }
        query?.let {
            for (i in origin_audios) {
                if (safeCheck(i.title, object : SafeCallCheckInt {
                        override fun check(): Boolean {
                            return i.title.lowercase(Locale.getDefault()).contains(
                                it.lowercase(Locale.getDefault())
                            )
                        }
                    })
                    || safeCheck(i.artist, object : SafeCallCheckInt {
                        override fun check(): Boolean {
                            return i.artist.lowercase(Locale.getDefault()).contains(
                                it.lowercase(Locale.getDefault())
                            )
                        }
                    }) || checkTittleArtists(i, it)
                ) {
                    audios.add(i)
                }
            }
        }
        setLoadingNow(false)
        view?.notifyListChanged()
    }

    fun fireQuery(q: String?) {
        query = if (q.isNullOrEmpty()) null else {
            q
        }
        updateCriteria()
    }

    public override fun onGuiResumed() {
        super.onGuiResumed()
        resolveRefreshingView()
        doAudioLoadTabs = if (doAudioLoadTabs) {
            return
        } else {
            true
        }
        view?.checkPermission()
    }

    private fun resolveRefreshingView() {
        resumedView?.displayRefreshing(
            loadingNow
        )
    }

    fun requestList() {
        setLoadingNow(true)
        if (bucket_id == 0) {
            audioListDisposable.add(Stores.instance
                .localMedia()
                .getAudios(accountId)
                .fromIOToMain()
                .subscribe({ onListReceived(it) }) { t ->
                    onListGetError(
                        t
                    )
                })
        } else {
            audioListDisposable.add(Stores.instance
                .localMedia()
                .getAudios(accountId, bucket_id.toLong())
                .fromIOToMain()
                .subscribe({ onListReceived(it) }) { t ->
                    onListGetError(
                        t
                    )
                })
        }
    }

    private fun onListReceived(data: List<Audio>) {
        if (data.isNullOrEmpty()) {
            actualReceived = true
            setLoadingNow(false)
            return
        }
        origin_audios.clear()
        actualReceived = true
        origin_audios.addAll(data)
        updateCriteria()
        setLoadingNow(false)
    }

    fun playAudio(context: Context, position: Int) {
        startForPlayList(context, audios, position, false)
        if (!Settings.get().other().isShow_mini_player) getPlayerPlace(accountId).tryOpenWith(
            context
        )
    }

    fun fireDelete(position: Int) {
        audios.removeAt(position)
        view?.notifyItemRemoved(position)
    }

    override fun onDestroyed() {
        audioListDisposable.dispose()
        super.onDestroyed()
    }

    fun fireRemoveClick(upload: Upload) {
        uploadManager.cancel(upload.id)
    }

    private fun onListGetError(t: Throwable) {
        setLoadingNow(false)
        showError(
            getCauseIfRuntime(t)
        )
    }

    fun fireFileForUploadSelected(file: String?) {
        val intent = UploadIntent(accountId, destination)
            .setAutoCommit(true)
            .setFileUri(Uri.parse(file))
        uploadManager.enqueue(listOf(intent))
    }

    fun fireFileForRemotePlaySelected(file: String?) {
        val intent = UploadIntent(accountId, remotePlay)
            .setAutoCommit(true)
            .setFileUri(Uri.parse(file))
        uploadManager.enqueue(listOf(intent))
    }

    fun getAudioPos(audio: Audio?): Int {
        if (!audios.isNullOrEmpty() && audio != null) {
            for ((pos, i) in audios.withIndex()) {
                if (i.id == audio.id && i.ownerId == audio.ownerId) {
                    i.isAnimationNow = true
                    view?.notifyItemChanged(
                        pos
                    )
                    return pos
                }
            }
        }
        return -1
    }

    fun firePermissionsCanceled() {
        errorPermissions = true
    }

    fun fireRefresh() {
        if (errorPermissions) {
            errorPermissions = false
            view?.checkPermission()
            return
        }
        audioListDisposable.clear()
        requestList()
    }

    fun fireScrollToEnd() {
        if (actualReceived) {
            requestList()
        }
    }

    override fun onGuiCreated(viewHost: IAudiosLocalView) {
        super.onGuiCreated(viewHost)
        viewHost.displayList(audios)
        viewHost.displayUploads(uploadsData)
        resolveUploadDataVisibility()
    }

    private fun onUploadsDataReceived(data: List<Upload>) {
        uploadsData.clear()
        uploadsData.addAll(data)
        resolveUploadDataVisibility()
    }

    private fun onUploadResults(pair: Pair<Upload, UploadResult<*>>) {
        val obj = pair.second.result as Audio
        if (obj.id == 0) view?.customToast?.showToastError(
            R.string.error
        )
        else {
            view?.customToast?.showToast(
                R.string.uploaded
            )
        }
    }

    private fun onProgressUpdates(updates: List<IProgressUpdate>) {
        for (update in updates) {
            val index = findIndexById(uploadsData, update.id)
            if (index != -1) {
                view?.notifyUploadProgressChanged(
                    index,
                    update.progress,
                    true
                )
            }
        }
    }

    private fun onUploadStatusUpdate(upload: Upload) {
        val index = findIndexById(uploadsData, upload.id)
        if (index != -1) {
            view?.notifyUploadItemChanged(
                index
            )
        }
    }

    private fun onUploadsAdded(added: List<Upload>) {
        for (u in added) {
            if (destination.compareTo(u.destination)) {
                val index = uploadsData.size
                uploadsData.add(u)
                view?.notifyUploadItemsAdded(
                    index,
                    1
                )
            }
        }
        resolveUploadDataVisibility()
    }

    private fun onUploadDeleted(ids: IntArray) {
        for (id in ids) {
            val index = findIndexById(uploadsData, id)
            if (index != -1) {
                uploadsData.removeAt(index)
                view?.notifyUploadItemRemoved(
                    index
                )
            }
        }
        resolveUploadDataVisibility()
    }

    private fun resolveUploadDataVisibility() {
        view?.setUploadDataVisible(uploadsData.isNotEmpty())
    }

    init {
        uploadsData = ArrayList(0)
        audios = ArrayList()
        origin_audios = ArrayList()
    }
}